spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ReactorLoadBalancerExchangeFilterFunction.java

```diff
@@ -0,0 +1,110 @@
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.client.loadbalancer.reactive;
+
+import java.net.URI;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.core.publisher.Mono;
+
+import org.springframework.cloud.client.ServiceInstance;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.reactive.function.client.ClientRequest;
+import org.springframework.web.reactive.function.client.ClientResponse;
+import org.springframework.web.reactive.function.client.ExchangeFilterFunction;
+import org.springframework.web.reactive.function.client.ExchangeFunction;
+
+/**
+ * An {@link ExchangeFilterFunction} that uses {@link ReactiveLoadBalancer} to execute
+ * requests against a correct {@link ServiceInstance}.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 2.2.0
+ */
+public class ReactorLoadBalancerExchangeFilterFunction implements ExchangeFilterFunction {
+
+	private static final Log LOG = LogFactory
+			.getLog(LoadBalancerExchangeFilterFunction.class);
+
+	private final ReactiveLoadBalancer.Factory<ServiceInstance> loadBalancerFactory;
+
+	public ReactorLoadBalancerExchangeFilterFunction(
+			ReactiveLoadBalancer.Factory<ServiceInstance> loadBalancerFactory) {
+		this.loadBalancerFactory = loadBalancerFactory;
+	}
+
+	@Override
+	public Mono<ClientResponse> filter(ClientRequest request, ExchangeFunction next) {
+		URI originalUrl = request.url();
+		String serviceId = originalUrl.getHost();
+		if (serviceId == null) {
+			String message = String.format(
+					"Request URI does not contain a valid hostname: %s",
+					originalUrl.toString());
+			if (LOG.isWarnEnabled()) {
+				LOG.warn(message);
+			}
+			return Mono.just(
+					ClientResponse.create(HttpStatus.BAD_REQUEST).body(message).build());
+		}
+		return choose(serviceId).map(response -> {
+			ServiceInstance instance = response.getServer();
+			if (instance == null) {
+				String message = serviceInstanceUnavailableMessage(serviceId);
+				if (LOG.isWarnEnabled()) {
+					LOG.warn(message);
+				}
+				throw new IllegalStateException(message);
+			}
+			else {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug(String.format(
+							"Load balancer has retrieved the instance for service %s: %s",
+							serviceId, instance.getUri()));
+				}
+				return instance;
+			}
+		}).flatMap(serviceInstance -> LoadBalancerUriTools.reconstructURI(serviceInstance,
+				originalUrl)).map(uri -> buildClientRequest(request, uri))
+				.flatMap(next::exchange)
+				.onErrorReturn(ClientResponse.create(HttpStatus.SERVICE_UNAVAILABLE)
```

Comment:
Why would you override the error with a client response here ? Shouldn't the error flow ?

Reply:
when we cannot get a valid instance, we want to get back to the user directly and respond with `SERVICE_UNAVAILABLE` so that they quickly know what has happened; we don't want to be going through with any other operations at this point.